File: src/main/java/de/fluxparticle/animation/object/ElementRectangle.java
package de.fluxparticle.animation.object;

import de.fluxparticle.animation.signal.Signal;
import javafx.beans.binding.DoubleExpression;
import javafx.geometry.BoundingBox;
import javafx.geometry.Bounds;
import javafx.scene.paint.Color;

public class ElementRectangle extends ElementNode {

	private final Signal<Double> x;
	public Signal<Double> xSignal() {
		return x;
	}

	private final Signal<Double> y;
	public Signal<Double> ySignal() {
		return y;
	}

	private final Signal<Double> width;
	public Signal<Double> widthSignal() {
		return width;
	}

	private final Signal<Double> height;
	public Signal<Double> heightSignal() {
		return height;
	}

	private final Signal<Color> fill;
	public Signal<Color> fillSignal() {
		return fill;
	}

	public ElementRectangle(DoubleExpression time, Bounds bs, Color fill) {
		super(time);
		this.x = new Signal<>(time, this, "x", bs.getMinX());
		this.y = new Signal<>(time, this, "y", bs.getMinY());
		this.width = new Signal<>(time, this, "w", bs.getWidth());
		this.height = new Signal<>(time, this, "h", bs.getHeight());
		this.fill = new Signal<>(time, this, "fill", fill);
	}

	@Override
	public <R> R accept(ElementNodeVisitor<R> visitor) {
		return visitor.visitRectangle(this);
	}

	public Bounds getBoundsInParent() {
		return new BoundingBox(x.lastValue(), y.lastValue(), width.lastValue(), height.lastValue());
	}

}

File: src/main/kotlin/de/fluxparticle/animation/object/ElementRectangle.kt
package de.fluxparticle.animation.`object`

import de.fluxparticle.animation.signal.Signal
import javafx.beans.binding.DoubleExpression
import javafx.geometry.BoundingBox
import javafx.geometry.Bounds
import javafx.scene.paint.Color

class ElementRectangle(time: DoubleExpression, bs: Bounds, fill: Color) : ElementNode(time) {

    private val x: Signal<Double>

    private val y: Signal<Double>

    private val width: Signal<Double>

    private val height: Signal<Double>

    private val fill: Signal<Color>

    val boundsInParent: Bounds
        get() = BoundingBox(x.lastValue(), y.lastValue(), width.lastValue(), height.lastValue())

    fun xSignal(): Signal<Double> {
        return x
    }

    fun ySignal(): Signal<Double> {
        return y
    }

    fun widthSignal(): Signal<Double> {
        return width
    }

    fun heightSignal(): Signal<Double> {
        return height
    }

    fun fillSignal(): Signal<Color> {
        return fill
    }

    init {
        this.x = Signal(time, this, "x", bs.minX)
        this.y = Signal(time, this, "y", bs.minY)
        this.width = Signal(time, this, "w", bs.width)
        this.height = Signal(time, this, "h", bs.height)
        this.fill = Signal(time, this, "fill", fill)
    }

    override fun <R> accept(visitor: ElementNodeVisitor<R>): R {
        return visitor.visitRectangle(this)
    }

}
